Add pkl input

psnairne opened the pull request.

psnairne committed:
The input of PMID_downloader is now a .pkl file

psnairne committed:
Got rid of an empty line

SmartMonkey-git committed:
Merge pull request #7 from VarenyaJ/get-PMID-pdf-files

PMID PDFs and the script

psnairne committed:
Merge branch 'main' into add-pkl-input

# Conflicts:
#	scripts/PMID_downloader.py
#	scripts/utils.py
#	tests/scripts/test_PMID_downloader.py

psnairne committed:
Lots to still to sort out here

psnairne committed:
Created a pmid_set to pkl script
Updated the readme
Updated test_PMID_downloader

Readme possibly a bit dodgy but I need to go home!

SmartMonkey-git commented on tests/scripts/test_PMID_downloader.py:
```diff
@@ -35,111 +40,115 @@ def pdf_bytes():
 
 @pytest.fixture()
-def pmids():
+def pmids_set():
```
If you want to make clear that this function returns/is a set. Pleas do:
```suggestion
def pmids() -> set:
```

SmartMonkey-git commented on scripts/utils.py:
```diff
@@ -57,3 +64,13 @@ def find_pmids(
     return pmids
+
+
+def pkl_to_set(pkl_file_path: str) -> set:
```
Function name is not consistent with what the function does, is it?

psnairne replied:
I assume you mean that it should be pkl_file_path_to_set :')

psnairne replied:
Oh I see. Very true

SmartMonkey-git commented on scripts/utils.py:
```diff
@@ -5,6 +5,13 @@
+import pickle
+
+
+def random_string() -> str:
+    return "".join(random.choice(string.ascii_letters) for _ in range(5))
```
This is not used anywhere.

psnairne replied:
I'll delete it (FYI I didn't write this function)

SmartMonkey-git replied:
I know, was fluted back into the project by merging.

SmartMonkey-git commented on tests/scripts/test_PMID_downloader.py:
```diff
@@ -35,111 +40,115 @@ def pdf_bytes():
+def test_PMID_downloader_no_pmcid_mocked(mock_entrez, pmids_no_pdf_set, request):
     """
-    This tests that everything works when we pass in PMIDs that do not correspond to PMCIDs.
+    This tests that everything works when we pass in a .pkl containing two PMIDs that do NOT correspond to PMCIDs.
```
Whats "everything"?

psnairne replied:
everything = all the code

SmartMonkey-git commented on scripts/create_pmid_pkl.py:
```diff
@@ -0,0 +1,36 @@
+"""
+)
+@click.argument("pmid_directory", type=click.Path(exists=True, dir_okay=True))
+@click.argument("pkl_file_path", type=click.Path(exists=False))
```
There should be a file_ok parameter or something in the path object and a dir_ok.

Think file_ok=True, while dir_ok=False

SmartMonkey-git commented on scripts/create_pmid_pkl.py:
```diff
@@ -0,0 +1,36 @@
+)
+@click.argument("pmid_directory", type=click.Path(exists=True, dir_okay=True))
+@click.argument("pkl_file_path", type=click.Path(exists=False))
+@click.option("--recursive_dir_search", is_flag=True, default=True)
```
What does is_flag do? Never used or seen it.

SmartMonkey-git commented on scripts/PMID_downloader.py:
```diff
@@ -91,29 +91,30 @@ def download_pdf(pmcid: str, pmid: str, pdf_out_dir: str):
         pdf_out_dir_path.mkdir(exist_ok=True, parents=True)
 
-    pmids = find_pmids(pmid_file_path, recursive=recursive_dir_search)
+    pmids: set = pkl_to_set(pkl_file_path)  # entries of the form "PMID_1234567"
```
Think, the comment here is rather a part of the doc string. Otherwise, you would need to mention that everywhere you call this function.

SmartMonkey-git commented on scripts/readme.md:
````diff
@@ -16,6 +16,21 @@ To filter the phenopackets by the cases use:
 python create_phenopacket_dataset.py "./data/tmp/cases" "./data/tmp/notebooks" "./data/tmp/phenopacket_dataset.csv" --recursive_ground_truth_dir True
 ```
 
+#### create_pmid_pkl script
````
Did you run the pipeline from scratch. Everything running fine?

psnairne replied:
No I need to test this! Will do it now.

SmartMonkey-git commented on scripts/readme.md:
````diff
@@ -16,6 +16,21 @@ To filter the phenopackets by the cases use:
+```shell
+python scripts/PMID_downloader.py "tmp/pmids.pkl" "data/tmp/pmid_pdfs"
+```
+This will download every PDF in the .pkl file to the directory data/tmp/pmid_pdfs whenever this is possible (i.e. whenever there is a valid PMCID).
````
There are no pdf's in the .pkl file.
Sorry, for being anal about it, but I think you can store bytes in a pkl, so this would be misleading.

SmartMonkey-git commented on tests/scripts/test_PMID_downloader.py:
```diff
@@ -35,111 +40,115 @@ def pdf_bytes():
-    This is an integration test, testing the selenium integration.
-    1. The first PMID corresponds to a PMCID and so will generate a PDF.
-    2. The second PMID does not correspond to a PMCID and so should not correspond to a PDF.
+    This tests that everything works when we pass in a .pkl file containing two PMIDS.
```
Try to no use first person singular or plural in docstrings.

Rather something like: "Tests that...."

psnairne replied:
Sure. Not used to writing code that might be looked at by other people (aka the public)

SmartMonkey-git commented on scripts/PMID_downloader.py:
```diff
@@ -91,29 +91,30 @@ def download_pdf(pmcid: str, pmid: str, pdf_out_dir: str):
-PMID_FILE_PATH:     Directory where PMID files are located.
-PDF_OUTPUT_DIR:     Ouput directory where PMID PDFs will be written.
+PKL_FILE_PATH:     the file path for the .pkl file
+PDF_OUTPUT_DIR:     where you want the directory containing the PDFs to be located
```
```suggestion
PDF_OUTPUT_DIR:     Directory to dump the PDFs
```

SmartMonkey-git commented on scripts/create_pmid_pkl.py:
```diff
@@ -0,0 +1,36 @@
+This takes a directory as an input, searches for strings of the form PMID_{1-8 digits} among the file names
+and writes those PMIDS to a .pkl file. 
+
+PMID_DIRECTORY:            The directory where you want to search for PMIDs among the file names.
```
```suggestion
PMID_DIRECTORY:            Directory to search for PMIDs among the file names.
```

SmartMonkey-git reviewed (commented):
Looks good! Think most of it is comment related. Would suggest all three of us have a look at: https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/

Lets agree, that within our code we only use the raw pmids i.e. the 1-8 digits and when storing a file, we add the PMID as a prefix. @VarenyaJ @psnairne
Please, keep that in mind. Might break things further down the line if we forget.

SmartMonkey-git commented on tests/scripts/test_create_pmid_pkl.py:
```diff
@@ -0,0 +1,38 @@
+from scripts.create_pmid_pkl import create_pmid_pkl
+
+
+@pytest.fixture()
+def pmids_set():
+    return {"PMID_8755636", "PMID_16636245"}
```
Nooo, only fixtures that are shared

psnairne replied:
As you wish

psnairne committed:
Improved based on Reuven's comments

psnairne committed:
Got the pipeline in readme.md working

Altered PMID_downloader so that you can specify how many PDFs you want to download at one time

psnairne committed:
Fixed the errors

psnairne committed:
Altered fixtures location

psnairne committed:
Formatting

SmartMonkey-git reviewed (approved):
Please, do the comments.
Otherwise it looks good!

Thank you! Your first PR, how are you feeling?

psnairne commented on scripts/PMID_downloader.py:
```diff
@@ -91,36 +90,57 @@ def download_pdf(pmcid: str, pmid: str, pdf_out_dir: str):
+        dl_batch_size = len(all_pmids)
+
+    if dl_batch_size > len(all_pmids):
+        click.secho(
+            message=f"Requested download batch size of {dl_batch_size} greater than number of PMIDs in {pkl_file_path}. Attempting to download all {len(all_pmids)} PMIDs.",
+            fg="yellow",
+        )
```
I still think it's worth having there. It's not necessarily a warning

psnairne commented on scripts/PMID_downloader.py:
```diff
@@ -91,36 +90,57 @@ def download_pdf(pmcid: str, pmid: str, pdf_out_dir: str):
 @click.argument("pdf_out_dir", type=click.Path(exists=False, dir_okay=True))
-@click.option("--recursive_dir_search", type=click.BOOL, default=True)
-def pmid_downloader(pmid_file_path: str, pdf_out_dir: str, recursive_dir_search: bool):
+@click.argument("dl_batch_size", type=int)
```
sure

psnairne commented on scripts/readme.md:
````diff
@@ -1,22 +1,37 @@
+```
+For every PMID in the .pkl file, a PDF will be downloaded to the directory *data/tmp/phenopacket_store/pmid_pdfs* whenever this is possible (i.e. whenever there is a valid PMCID). 
+
+### File to Phenopacket Script
````
Why not?

psnairne committed:
Took into account some of Reuven's suggestions

Updated the scripts readme.md to reflect my understanding of the pipeline

psnairne committed:
It's any not Any

VarenyaJ commented on scripts/PMID_downloader.py:
```diff
@@ -91,36 +90,57 @@ def download_pdf(pmcid: str, pmid: str, pdf_out_dir: str):
         pdf_out_dir_path.mkdir(exist_ok=True, parents=True)
 
-    pmids = find_pmids(pmid_file_path, recursive=recursive_dir_search)
+    all_pmids: set = pkl_loader(pkl_file_path)
+
+    if dl_cut_off == 0:
+        dl_cut_off = len(all_pmids)
```
What happens if there is a valid `.pkl` file but it is empty?

psnairne commented on tests/scripts/test_PMID_downloader.py:
```diff
@@ -35,111 +38,110 @@ def pdf_bytes():
         for pdf in pdf_file_names:
             # expected PDF has file size ≈ 204,000 bytes
-            assert os.path.getsize(f"{tmp_dir}/{pdf}") >= 200000
+            assert os.path.getsize(f"{output_dir}/{pdf}") >= 200000
```
sure

psnairne committed:
created variable for min valid pdf bytes
